### PR TITLE
Refactor expressions

### PR DESCRIPTION
**Pull request by sundy-li**
## Summary
RFC: https://github.com/sundy-li/fuse-query/blob/action_dag/website/datafuse/docs/rfcs/query/2021-05-22-plan-expression-refactor.md

- [x] split ScalarFunction and AggregateFunction
- [x] expression rewriter and expression visitor
- [x] expression chain in AST to plan builder (simple version as ActionsDag)
- [x] testing

### A complex Example

```
EXPLAIN actions = 1
SELECT
    number + 1 AS number,
    number + 2 AS a,
    sum(number % 3) + 4 AS b
FROM numbers(10)
GROUP BY
    number,
    a
HAVING (b + 1) > 3

Query id: 9675ed9e-81fc-40dc-b9c7-5c33746db915

┌─explain───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Expression (Projection)                                                                                                                                   │
│ Actions: PROJECT plus(number, 1) AS number, plus(plus(number, 1), 2) AS a, plus(sum(modulo(plus(number, 1), 3)), 4) AS b                                  │
│   Expression (Before ORDER BY and SELECT)                                                                                                                 │
│   Actions: PROJECT plus(number, 1), plus(plus(number, 1), 2), plus(sum(modulo(plus(number, 1), 3)), 4)                                                    │
│     Filter (HAVING)                                                                                                                                       │
│     Filter column: greater(plus(plus(sum(modulo(plus(number, 1), 3)), 4), 1), 3)                                                                          │
│     Actions: PROJECT plus(number, 1), plus(plus(number, 1), 2), sum(modulo(plus(number, 1), 3))                                                           │
│              ADD 4 UInt8 Const(UInt8)                                                                                                                     │
│              ADD 1 UInt8 Const(UInt8)                                                                                                                     │
│              ADD 3 UInt8 Const(UInt8)                                                                                                                     │
│              FUNCTION plus(sum(modulo(plus(number, 1), 3)), 4) UInt64 = plus(sum(modulo(plus(number, 1), 3)), 4)                                          │
│              REMOVE sum(modulo(plus(number, 1), 3))                                                                                                       │
│              REMOVE 4                                                                                                                                     │
│              FUNCTION plus(plus(sum(modulo(plus(number, 1), 3)), 4), 1) UInt64 = plus(plus(sum(modulo(plus(number, 1), 3)), 4), 1)                        │
│              REMOVE 1                                                                                                                                     │
│              FUNCTION greater(plus(plus(sum(modulo(plus(number, 1), 3)), 4), 1), 3) UInt8 = greater(plus(plus(sum(modulo(plus(number, 1), 3)), 4), 1), 3) │
│              REMOVE plus(plus(sum(modulo(plus(number, 1), 3)), 4), 1)                                                                                     │
│              REMOVE 3                                                                                                                                     │
│       Aggregating                                                                                                                                         │
│       Keys: plus(number, 1), plus(plus(number, 1), 2)                                                                                                     │
│       Aggregates:                                                                                                                                         │
│           sum(modulo(plus(number, 1), 3))                                                                                                                 │
│             Function: sum(UInt8) → UInt64                                                                                                                 │
│             Arguments: modulo(plus(number, 1), 3)                                                                                                         │
│             Argument positions: 2                                                                                                                         │
│         Expression (Before GROUP BY)                                                                                                                      │
│         Actions: ADD 1 UInt8 Const(UInt8)                                                                                                                 │
│                  ADD 2 UInt8 Const(UInt8)                                                                                                                 │
│                  ADD 3 UInt8 Const(UInt8)                                                                                                                 │
│                  FUNCTION plus(number, 1) UInt64 = plus(number, 1)                                                                                        │
│                  REMOVE number                                                                                                                            │
│                  REMOVE 1                                                                                                                                 │
│                  FUNCTION plus(plus(number, 1), 2) UInt64 = plus(plus(number, 1), 2)                                                                      │
│                  REMOVE 2                                                                                                                                 │
│                  FUNCTION modulo(plus(number, 1), 3) UInt8 = modulo(plus(number, 1), 3)                                                                   │
│                  REMOVE 3                                                                                                                                 │
│           SettingQuotaAndLimits (Set limits and quota after reading from storage)                                                                         │
│             ReadFromStorage (SystemNumbers)                                                                                                               │
└───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

38 rows in set. Elapsed: 0.014 sec.

```

There are some main steps to go:

1. Alias extracting and expression rewritten by aliases, results will be 

```
SELECT
    number + 1 AS number,
    (number + 1) + 2 AS a,
    sum((number+1)% 3) + 4  as b
FROM numbers(10) group by (number + 1), (number + 1) + 2  having (sum((number+1) % 3) + 4) + 1 > 3
```

2. Apply expressions before group by, visit all expressions except aggregate functions.

eg expressions:
```
plus(number, 1)
plus(plus(number, 1), 2)
modulo(plus(number, 1), 3)
```

3. If we already applied expressions, we can make expressions materialized as a Column expression for downstream to get from the block.

```
ExpressionFunction :  plus(number, 1)     ---> ExpressionColumn:  plus(number, 1) 
```

3. Apply expressions to the group by,  visit all expressions to find aggregate functions.

```
 Keys: plus(number, 1), plus(plus(number, 1), 2)
 Aggregates:
    sum(modulo(plus(number, 1), 3))   
```

4. Apply expressions after group by

eg: 
```
FUNCTION plus(sum(modulo(plus(number, 1), 3)), 4)
```

## Changelog

- Improvement

### CODE REVIEW

File: website/datafuse/docs/rfcs/query/2021-05-22-plan-expression-refactor.md [hunk 1]
```diff
@@ -0,0 +1,166 @@
+# Expression and plan builder refactoring docs
+
+
+## Summary
+
+Logic plan and expression play a big role throughout the life cycle of SQL query. This doc is intended to explain the new design of expressions and plan builder.
+
+> Let's introduce expressions firstyly
```
Comment (jyizheng) [inline website/datafuse/docs/rfcs/query/2021-05-22-plan-expression-refactor.md:8]: typo: firstly

File: common/planners/src/plan_builder_test.rs [hunk 2]
```diff
@@ -2,17 +2,20 @@
...
     use pretty_assertions::assert_eq;
 
-    use crate::*;
-
     struct TestCase {
         name: &'static str,
         plan: Result<PlanNode>,
-        expect: &'static str
+        expect: &'static str,
+        err: &'static str
```
Comment (jyizheng) [inline common/planners/src/plan_builder_test.rs:18]: It is neat to add the expected error here.

File: common/aggregate_functions/src/aggregate_avg.rs [hunk 3]
```diff
@@ -4,53 +4,41 @@
...
 use common_datavalues::DataType;
 use common_datavalues::DataValue;
 use common_datavalues::DataValueAggregateOperator;
 use common_datavalues::DataValueArithmetic;
 use common_datavalues::DataValueArithmeticOperator;
-use common_exception::ErrorCodes;
 use common_exception::Result;
 
-use crate::IFunction;
+use crate::IAggreagteFunction;
```
Comment (jyizheng) [inline common/aggregate_functions/src/aggregate_avg.rs:17]: Do you mean "IAggregateFunction"?

Comment (sundy-li) [reply common/aggregate_functions/src/aggregate_avg.rs:17]: Oh, typo

File: common/planners/src/plan_builder_test.rs (hunk 2 above)
Comment (sundy-li) [reply common/planners/src/plan_builder_test.rs:18]: err is expected error msg, empty str for normal return.

File: common/aggregate_functions/src/aggregate_function_factory.rs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+        map
+    };
+}
+
+impl AggregateFunctionFactory {
+    pub fn get(name: &str) -> Result<Box<dyn IAggregateFunction>> {
+        let map = FACTORY.read();
+        let creator = map.get(&*name.to_lowercase()).ok_or_else(|| {
+            ErrorCodes::UnknownFunction(format!("Unsupported AggregateFunction: {}", name))
```
Comment (zhang2014) [inline common/aggregate_functions/src/aggregate_function_factory.rs:34]: UnknownAggregateFunction is better?

File: common/aggregate_functions/src/test.rs [hunk 5]
```diff
@@ -0,0 +1,3 @@
+// Copyright 2020-2021 The Datafuse Authors.
+//
+// SPDX-License-Identifier: Apache-2.0.
```
Comment (zhang2014) [inline common/aggregate_functions/src/test.rs:3]: maybe remove it?

File: common/functions/src/test.rs [hunk 6]
```diff
@@ -0,0 +1,3 @@
+// Copyright 2020-2021 The Datafuse Authors.
+//
+// SPDX-License-Identifier: Apache-2.0.
```
Comment (zhang2014) [inline common/functions/src/test.rs:3]: remove it?

File: common/planners/src/plan_display.rs [hunk 7]
```diff
@@ -62,7 +62,7 @@ impl PlanNode {
                                     f,
                                     "{:?}:{:?}",
                                     plan.exprs[i],
-                                    plan.schema().fields()[i].data_type()
+                                    plan.exprs[i].to_data_type(&plan.input.schema()).unwrap()
```
Comment (zhang2014) [inline common/planners/src/plan_display.rs:65]: `unwrap` can cause panic. use `?` operator

Comment (sundy-li) [reply common/planners/src/plan_display.rs:65]: It's for display, may need ErrorCodes to fmt error

File: common/planners/src/plan_expression_action_test.rs [hunk 8]
```diff
@@ -1,15 +1,16 @@
+use std::sync::Arc;
+
+use pretty_assertions::assert_eq;
+
+use crate::test::Test;
+use crate::*;
+
 // Copyright 2020-2021 The Datafuse Authors.
```
Comment (zhang2014) [inline common/planners/src/plan_expression_action_test.rs:8]: Should be move to file header

File: fusequery/query/src/pipelines/transforms/transform_expression_executor.rs [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+use common_exception::ErrorCodes;
+use common_exception::Result;
+use common_planners::ActionNode;
+use common_planners::ExpressionAction;
+use common_planners::ExpressionChain;
+
+/// ExpressionExecutor is a helper struct for expressions and projections
+/// Aggregate functions is not covered, because all expressions in aggregate functions functions are executed.
+#[derive(Debug, Clone)]
+pub struct ExpressionExecutor {
```
Comment (zhang2014) [inline fusequery/query/src/pipelines/transforms/transform_expression_executor.rs:20]: 👍 Great work.

**Review by zhang2014 — APPROVED** [review]
LGTM

**Comment by ives9638** [comment]
Feat!

**Pull request merged**